"Element with presentational children has no focusable content" (rule 307n5z) - adding examples

dan-tripp-siteimprove opened the pull request:
Adding examples to rule presentational-children-no-focusable-content-307n5z.  Mostly uncontroversial, I think, except for Inapplicable Example 4 ("At the time of writing..."), which I expect will ruffle some feathers.

Closes issue(s): none 

Need for Call for Review:
This will not require a Call for Review : editorial changes (including to the applicability, expectation or examples section)



---

## Pull Request Etiquette

### **When creating PR:**

- [ ] Make sure you're requesting to **pull a branch** (right side) to the `develop` branch (left side).
- [ ] Make sure you do not remove the "How to Review and Approve" section in your pull request description

### **After creating PR:**

- [ ] Add yourself (and co-authors) as "Assignees" for PR.
- [ ] Add label to indicate if it's a `Rule`, `Definition` or `Chore`.
- [ ] [Link the PR to any issue it solves](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue). This will be done automatically by [referencing the issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) at the top of this comment in the indicated place.
- [ ] Optionally request feedback from anyone in particular by assigning them as "Reviewers".

### **When merging a PR:**

- [ ] Close any issue that the PR resolves. This will happen automatically upon merging if the PR was correctly linked to the issue, e.g. by referencing the issue at the top of this comment.

## How to Review And Approve

- Go to the “Files changed” tab
- Here you will have the option to leave comments on different lines.
- Once the review is completed, find the “Review changes” button in the top right, select “Approve” (if you are really confident in the rule) or "Request changes" and click “Submit review”.
- Make sure to also review the proposed Call for Review period. In case of disagreement, the longer period wins.

dan-tripp-siteimprove committed:
invalid-form-field-value-36b590: updating failed example 2.

Jym77 committed:
Merge branch 'develop' into develop

dan-tripp-siteimprove committed:
Merge branch 'act-rules:develop' into develop

dan-tripp-siteimprove committed:
Merge branch 'act-rules:develop' into develop

dan-tripp-siteimprove committed:
Merge branch 'act-rules:develop' into develop

dan-tripp-siteimprove committed:
Adding examples to rule presentational-children-no-focusable-content-307n5z

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,16 +136,93 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
 #### Inapplicable Example 1
 
-This element has a `link` role which does not have [presentational children][].
+None of the roles involved in this semantic table have [presentational children][].  (The roles are `table`, `row`, and `cell`, respectively, for the `<table>`, `<tr>`, and `<td>` elements.)  So this rule does not apply.
```
There are actually also a `button` and a `link` role in the example, so the list of roles is not correct.

dan-tripp-siteimprove replied:
```suggestion
None of the roles involved in this semantic table have [presentational children][].  (The roles are `table`, `row`, `cell`, `button`, and `link`, for the `<table>`, `<tr>`, `<td>`, `<button>`, and `<a>` elements, respectively.)  So this rule does not apply.
```

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,16 +136,93 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
+</li>
+</ul>
+
+Adding to this confusion is a third term: "decorative".  The words "decorative" and "presentational" are often used interchangeably, but that usage is inaccurate.  The word "decorative" often appears in a sentence such as "marking an image as decorative" - that is, by adding `alt=""` to an `<img>` element.  "Decorative" in that context <i>does</i> mean "hidden" - and "hidden", again, is different ...
```
That's why we use "**marked** as decorative", and not "decorative" 😄 
Being "marked as decorative" is showing author **intention** to somehow remove the content. It doesn't necessarily means that the content will effectively be removed/hidden/… (because the presentational role conflict resolution can trigger).

I also consider "decorative" as being a semantic property of the page. "presentational" (or role of `presentation` (post-conflict resolution)) is more of a functional/effective property of the page. The same way that "being a label (WCAG sense of the term)" is a semantic property while using a `label` element is the functional property (and the way to convey  that semantics programmatically). That is, an image is "decorative" not because it has `alt=""` but because it is "serving only an aesthetic purpose, providing no information, and having no functionality" (WCAG's "pure decoration"). Adding `alt=""` is a way to convey programmatically the fact that the image is decorative, so that ATs can treat it differently from a non-decorative image.

In that spirit, "marked as decorative" is "the author tried to convey programmatically the semantic information that something is decorative". This may work as intended (and the stuff will be either presentational or hidden), but this may also fail (and the stuff will be exposed as normal).

dan-tripp-siteimprove replied:
Okay, so decorative != marked as decorative.  Even if that's okay (which is unclear to me), I still think we have a problem.  

I would draw a venn diagram here if I could, but I can't, so here is a tree which represents a venn diagram of the situation you described.  (In this venn diagram tree, all child sets totally consume the parent set.  eg. Set B minus Set C minus Set D = an empty set.)

- Set A: marked as decorative 
  - Set B: "marked as decorative" worked 
    - Set C: presentational 
    - Set D: hidden
  - Set E: "marked as decorative" did not work (==> still has semantics) 

That presents two problems:
1) Sets C and D are so different, I think it's confusing for us to include them both under "Set B: "marked as decorative" worked".
2) I think that "Set D: hidden" corresponds to what most people think when they hear "marked as decorative".  Also, I suspect that it overlaps precisely with their ideas of "decorative" and WCAG's "pure decoration" - and that's a good thing.  If we go with the the above venn diagram, someone might say that ACT is out of touch, and they would be right.

carlosapaduarte replied:
While I think this feedback about how some definitions can be perceived is important, I don't think this is a good place to discuss it. I would suggest @dan-tripp-siteimprove opens an issue to discuss these points, and rewrite the example description is a less verbose manner.

Jym77 reviewed (commented):
The examples are good. The descriptions are very verbose and going into much more details compared to what we usually do. Some of these discussions might be moved to the background section instead (e.g. the discussion about presentational/hidden/decorative.

dan-tripp-siteimprove committed:
Update _rules/presentational-children-no-focusable-content-307n5z.md

dan-tripp-siteimprove commented:
> Some of these discussions might be moved to the background section instead (e.g. the discussion about presentational/hidden/decorative.

I can do that.  How about I move Inapplicable Examples 3 and 4 to the Background section?

carlosapaduarte commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,16 +136,93 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
+
+#### Failed Example 5
+
+This element with the `img` role contains an `<a>` element.  The `img` role has [presentational children][].  The `<a>` element is included in [sequential focus navigation][].  So the element with the `img` role fails the rule.  (Interestingly, no similar failure is possible with a standard HTML `<img>` element, because according to the HTML specification, the `<img>` element has no closing tag.  ...
```
```suggestion
This element with the `img` role contains an `<a>` element.  The `img` role has [presentational children][].  The `<a>` element is included in [sequential focus navigation][].  So the element with the `img` role fails the rule.
```

carlosapaduarte commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,16 +136,93 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
 
+#### Inapplicable Example 2
+
+This `<a>` element has a `link` role, which does not have [presentational children][].  So this element - perhaps surprisingly - does not fail this rule, because it's inapplicable.  This is the case even though it's invalid HTML.  Any time there's a link as the <i>outer</i> element, that will be inapplicable under this rule.  (If the elements were switched - that is, if the `<button>` element con ...
```
```suggestion
This `<a>` element has a `link` role, which does not have [presentational children][].  So this element does not fail this rule, because it's inapplicable. 
```

carlosapaduarte commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,16 +136,93 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
+
+#### Inapplicable Example 3
+
+The applicability of this rule does not include elements with an <i>explicit</i> role of `presentation`.  Only <i>implicit</i> (through roles with [presentational children][]).  This example shows that. 
```
```suggestion
The applicability of this rule does not include elements with an <i>explicit</i> role of `presentation`.  Only <i>implicit</i>, through roles with [presentational children][]. 
```

carlosapaduarte reviewed (dismissed):
Tried to cut down on the verbosity of the example's descriptions.

dan-tripp-siteimprove committed:
Update _rules/presentational-children-no-focusable-content-307n5z.md

Co-authored-by: Carlos Duarte <carlosapaduarte@gmail.com>

dan-tripp-siteimprove committed:
Update _rules/presentational-children-no-focusable-content-307n5z.md

Co-authored-by: Carlos Duarte <carlosapaduarte@gmail.com>

dan-tripp-siteimprove committed:
Update _rules/presentational-children-no-focusable-content-307n5z.md

Co-authored-by: Carlos Duarte <carlosapaduarte@gmail.com>

dan-tripp-siteimprove committed:
Inapplicable Example 4:
- reducing verbosity.
- removing most commentary on "hidden" vs. "presentational".
- removing all commentary on "marked as decorative".

dan-tripp-siteimprove commented:
> I would suggest @dan-tripp-siteimprove opens an issue to discuss these points, and rewrite the example description is a less verbose manner.
@carlosapaduarte  That's fair.  I haven't opened a new issue yet - I'll discuss it elsewhere first.  I did rewrite the example description and pushed another commit (d161d08ba6eba31686eee65f35f54bc425725e79).  Let me know what you think.

dan-tripp-siteimprove committed:
Updating Inapplicable Example 4 and references to rule 6cfa84.

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,16 +135,79 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
 #### Inapplicable Example 1
 
-This element has a `link` role which does not have [presentational children][].
+None of the roles involved in this semantic table have [presentational children][].  (The roles are `table`, `row`, `cell`, `button`, and `link`, for the `<table>`, `<tr>`, `<td>`, `<button>`, and `<a>` elements, respectively.)  So this rule does not apply.
```
Wouldn't the rule apply to the `button`? (and pass because it has no focusable child).

dan-tripp-siteimprove replied:
Yes.  I muddled this example. I was trying to get across the idea that the `<table>`, `<tr>`, and `<td>` elements (only) are inapplicable.  I just pushed a commit (9dd1ac652891bf99d4d7ea1f6c8598ccdb7c092a) that tries to make it more clear.

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,16 +135,79 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
 
+#### Inapplicable Example 2
+
+This `<a>` element has a `link` role, which does not have [presentational children][].  So this element does not fail this rule, because it's inapplicable. 
```
Same here, the rule would be Applicable to the `button` (only), and pass.

dan-tripp-siteimprove replied:
Fair point.  My intent was to comment on the inapplicability of the `<a>` element - the `<button>` element was just there as an example of a focusable child element.  I just pushed 527cfd96cec86da0fa5850834c83cc1691302482 which, I hope, clears it up.

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
````diff
@@ -127,16 +135,79 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
+		<a href="#">inapplicable</a>
+	</li>
+</ul>
+```
````
As per our consistency rules, we need to make sure that the Inapplicable examples do not fail the accessibility requirement (in this case, 4.1.2). I'm not sure exactly what happens in that case (and whether the presentational conflict triggers or not).

dan-tripp-siteimprove replied:
Good point.  I read [test-cases.md](https://github.com/act-rules/act-rules.github.io/blob/d703022a0298f1996f980ee204b5304fa9aecaf4/pages/design/test-cases.md?plain=1).  Its guidance of "Consistency With Accessibility Requirements", combined with [Presentational Roles Conflict Resolution](https://www.w3.org/TR/wai-aria-1.2/#conflict_resolution_presentation_none) make it so that I don't know how to create an inapplicable example which explains this aspect of the rule.

I have seen customers wrongly report a false positive for this rule, and they write something like: "This is a false positive because this element does not have role=presentation.  It has role=tab."   Then I reply: "This rule isn't about an explicit role of presentation.  It's about an /implicit/ role of presentation."  So with this inapplicable example, I was trying to help explain that.  

So I tried something different just now in commit 9f81ae4a95da9ae5b412ccb31199917d9cbef491: I removed this inapplicable example and added some equivalent wording in the "background" section.  What do you think?

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
````diff
@@ -127,16 +135,79 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
+<button aria-hidden="true">
+	<a href="#">inapplicable</a>
+</button>
+```
````
Same here. Dose that pass 4.1.2?
Also, `aria-hidden` is actually ignored on focusable stuff as per [Including Elements in the Accessibility Tree](https://www.w3.org/TR/wai-aria-1.2/#tree_inclusion): 
> user agents MUST provide an [accessible object](https://www.w3.org/TR/wai-aria-1.2/#dfn-accessible-object) in the [accessibility tree](https://www.w3.org/TR/wai-aria-1.2/#dfn-accessibility-tree) for DOM [elements](https://www.w3.org/TR/wai-aria-1.2/#dfn-element) that meet any of the following criteria:
> 
> * Elements that are not [hidden](https://www.w3.org/TR/wai-aria-1.2/#dfn-hidden) and may fire an [accessibility API](https://www.w3.org/TR/wai-aria-1.2/#dfn-accessibility-api) [event](https://www.w3.org/TR/wai-aria-1.2/#dfn-event), including:
>   * Elements that are currently focused, even if the element or one of its ancestor elements has its [aria-hidden](https://www.w3.org/TR/wai-aria-1.2/#aria-hidden) attribute set to true.

So, I'm not sure how this ends up ultimately.

dan-tripp-siteimprove replied:
For this one too, I replaced this example with some wording in the "background" section.  I did this in commit c8d188d7589b27abbcc7793b356bc04e60c0d2ed.  Let me know what you think.

dan-tripp-siteimprove committed:
Updating Inapplicable Example 1 because of https://github.com/act-rules/act-rules.github.io/pull/2099#discussion_r1410407851

dan-tripp-siteimprove committed:
updating Inapplicable Example 2 based on https://github.com/act-rules/act-rules.github.io/pull/2099#discussion_r1410408598

dan-tripp-siteimprove committed:
removing Inapplicable Example 3 and replacing it w/ wording in the "background" section.  in response to https://github.com/act-rules/act-rules.github.io/pull/2099#discussion_r1410411987

dan-tripp-siteimprove committed:
updating Background in response to https://github.com/act-rules/act-rules.github.io/pull/2099#discussion_r1410416038

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,14 +137,54 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
 #### Inapplicable Example 1
 
-This element has a `link` role which does not have [presentational children][].
+The roles that build this semantic table structure (`table` for `<table>`, `row` for `<tr>`, and `cell` for `<td>`) do not have [presentational children][].  So this rule does not apply to them.  (This rule does apply to the `<button>` element within the table.  The `<button>` element passes the rule.)
```
I still believe the `button` should go away…
From an implementer point of view, tools will report "Passed" rather than "Inapplicable" on that example. This is not a huge deal as it they are equivalent for consistency, but it still creates some weirdness. 

We could have this example as a Passed Example. It would still fulfil the same goal (show that `table` role and its friends are not applicable) (because tools that apply the rule to them would report a Failed outcome). OTOH, I am not sure that the `button` is adding much value to the example (the link is already a focusable element inside the table, we can have two links and the example is still working as intended).

dan-tripp-siteimprove replied:
Fair points, I'll discuss one-on-one some time.

carlosapaduarte replied:
I agree with @Jym77 

If we want this as an inapplicable example then we should not ask implementations to output "Passed". I agree with either solution proposed: make this a passed example or remove the `button` element.

dan-tripp-siteimprove replied:
I fixed this a minute ago by removing the button element.

carlosapaduarte committed:
Merge branch 'develop' into presentational-children-no-focusable-content-307n5z---examples-aug-2023

dan-tripp-siteimprove committed:
Updating "Inapplicable Example 1" as per https://github.com/act-rules/act-rules.github.io/pull/2099/files#r1426483648

dan-tripp-siteimprove committed:
Inapplicable Example 1: fixing whitespace.

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -42,16 +42,17 @@ This rule applies to elements with a [semantic role][] that defines its [childre
 
 Elements with a [semantic role][] that has [presentational children][] will not have any descendants in the accessibility tree. If any of those descendants are included in [sequential focus navigation][], this causes the focus to land on an element that has no corresponding node in the [accessibility tree][]. The result is that there is no programmatic name or role available for assistive technol ...
 
+This rule is often misunderstood as applying to elements with an <i>explicit</i> role of `presentation`.  In fact, this rule only applies to elements which have been given an <i>implicit</i> role of `presentation` through the [presentational children][] mechanism.  Similarly, this rule does not apply to elements with `aria-hidden="true"`.
```
```suggestion
This rule is often misunderstood as applying to elements with an _explicit_ role of `presentation`.  In fact, this rule only applies to elements which have been given an _implicit_ role of `presentation` through the [presentational children][] mechanism.  Similarly, this rule does not apply to elements with `aria-hidden="true"`.
```
(which Markdown  conversion then translates to `em` elements)

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -93,6 +94,15 @@ This element with `menuitemcheckbox` role has an `input` element as a descendant
 
+#### Passed Example 4
+
+This `<button>` element has an `<a>` element as a [child][].  The `<a>` element has no `href` attribute, so it isn't included in [sequential focus navigation][].  So this `<button>` element passes the rule.
```
```suggestion
This `<button>` element has an `a` element as a [child][].  The `a` element has no `href` attribute, so it isn't included in [sequential focus navigation][].  So this `button` element passes the rule.
```
We don't use angled brackets for elements' names in ACT Rules…

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,14 +137,53 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
 
+#### Failed Example 4
+
+This element with the `tab` role contains an `<a>` element.  The `tab` role has [presentational children][].  The `a` element is included in [sequential focus navigation][].  So the element with the `tab` role fails the rule.  (This tablist implementation is non-functional for users.  It's not meant to function - it's only meant to show roles.)
```
```suggestion
This element with the `tab` role contains an `a` element.  The `tab` role has [presentational children][].  The `a` element is included in [sequential focus navigation][].  So the element with the `tab` role fails the rule.  (This tablist implementation is non-functional for users.  It's not meant to function - it's only meant to show roles.)
```

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,14 +137,53 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
+
+#### Failed Example 5
+
+This element with the `img` role contains an `<a>` element.  The `img` role has [presentational children][].  The `<a>` element is included in [sequential focus navigation][].  So the element with the `img` role fails the rule.
```
```suggestion
This element with the `img` role contains an `a` element.  The `img` role has [presentational children][].  The `a` element is included in [sequential focus navigation][].  So the element with the `img` role fails the rule.
```

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,14 +137,53 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
 #### Inapplicable Example 1
 
-This element has a `link` role which does not have [presentational children][].
+None of the roles that build this semantic table structure (`table` for `<table>`, `row` for `<tr>`, `columnheader` for `<th>`, and `cell` for `<td>`) have [presentational children][].  So this rule does not apply to them.
```
```suggestion
None of the roles that build this semantic table structure (`table` for `table`, `row` for `tr`, `columnheader` for `th`, and `cell` for `td`) have [presentational children][].  So this rule does not apply to them.
```

Jym77 commented on _rules/presentational-children-no-focusable-content-307n5z.md:
```diff
@@ -127,14 +137,53 @@ This element with `menuitemcheckbox` role has a checkbox as a child. Because the
+
+#### Inapplicable Example 2
+
+This `<a>` element has a `link` role, which does not have [presentational children][].  So this `<a>` element does not fail this rule, because it's inapplicable.  To have a "focusable element within a focusable element" like this is a bad practice, but this rule doesn't directly check for it.
```
```suggestion
This `a` element has a `link` role, which does not have [presentational children][].  So this `a` element does not fail this rule, because it's inapplicable.  To have a "focusable element within a focusable element" like this is a bad practice, but this rule doesn't directly check for it.
```

Jym77 reviewed (changes requested):
Just some cleanup.

dan-tripp-siteimprove committed:
Update _rules/presentational-children-no-focusable-content-307n5z.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

dan-tripp-siteimprove committed:
Update _rules/presentational-children-no-focusable-content-307n5z.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

dan-tripp-siteimprove committed:
Update _rules/presentational-children-no-focusable-content-307n5z.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

dan-tripp-siteimprove committed:
Update _rules/presentational-children-no-focusable-content-307n5z.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

dan-tripp-siteimprove committed:
Update _rules/presentational-children-no-focusable-content-307n5z.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

dan-tripp-siteimprove committed:
Update _rules/presentational-children-no-focusable-content-307n5z.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

Jym77 committed:
Merge branch 'develop' into presentational-children-no-focusable-content-307n5z---examples-aug-2023
